### PR TITLE
Introduce basic "cudf" backend for Dask Expressions

### PR DESCRIPTION
**Pull request by rjzamora**
## Description

Mostly addresses https://github.com/rapidsai/cudf/issues/15027

https://github.com/dask-contrib/dask-expr/pull/728 exposed the necessary mechanisms for us to define a custom dask-expr backend for `cudf`. The new dispatching mechanisms are effectively the same as those in `dask.dataframe`. The only difference is that we are now registering/implementing "expression-based" collections.

This PR does the following:
- Defines a basic `DataFrameBackendEntrypoint` class for collection creation, and registers new collections using `get_collection_type`.
- Refactors the `dask_cudf` import structure to properly support the `"dataframe.query-planning"` configuration.
- Modifies CI to test dask-expr support for some of the `dask_cudf` tests. This coverage can be expanded in follow-up work.

~**Experimental Change**: This PR patches `dask_expr._expr.Expr.__new__` to enable type-based dispatching. This effectively allows us to surgically replace problematic `Expr` subclasses that do not work for cudf-backed data. For example, this PR replaces the upstream `TakeLast` expression to avoid using `squeeze` (since this method is not supported by cudf). This particular fix can be moved upstream relatively easily. However, having this kind of "patching" mechanism may be valuable for more complicated pandas/cudf discrepancies.~

## Usage example

```python
from dask import config
config.set({"dataframe.query-planning": True})
import dask_cudf

df = dask_cudf.DataFrame.from_dict(
    {"x": range(100), "y":  [1, 2, 3, 4] * 25, "z": ["1", "2"] * 50},
    npartitions=10,
)
df["y2"] = df["x"] + df["y"]
agg = df.groupby("y").agg({"y2": "mean"})["y2"]
agg.simplify().pprint()
```
Dask cuDF should now be using dask-expr for "query planning":
```
Projection: columns='y2'
  GroupbyAggregation: arg={'y2': 'mean'} observed=True split_out=1'y'
    Assign: y2=
      Projection: columns=['y']
        FromPandas: frame='<dataframe>' npartitions=10 columns=['x', 'y']
      Add:
        Projection: columns='x'
          FromPandas: frame='<dataframe>' npartitions=10 columns=['x', 'y']
        Projection: columns='y'
          FromPandas: frame='<dataframe>' npartitions=10 columns=['x', 'y']
```

## TODO

- [x] Add basic tests
- [x] Confirm that general design makes sense

**Follow Up Work**:

- Expand dask-expr test coverage
- Fix local and upstream bugs
- Add documentation once "critical mass" is reached

### CODE REVIEW

File: dependencies.yaml
Comment (rjzamora) [inline dependencies.yaml:667]: @charlesbluca - Sorry for the ping, but I thought you might have advice :)

I'd like to start running a subset of dask_cudf tests with dask_expr installed (and "enabled" within `dask.dataframe`).  Do you know what kind of hoops I need to jump through to add a new dependency for dask_cudf testing these days?

Comment (charlesbluca) [reply dependencies.yaml:667]: This is a good start, but IIUC think we'll want to split this dependency out because the way it would be specified between conda env files and requirements.txt / pyproject.toml is different; think something like this should work:

```yaml
  test_python_dask_cudf:
    common:
      - output_types: [conda, requirements, pyproject]
        packages:
          - dask-cuda==24.4.*
          - *numba
      - output_types: conda
        packages:
          - pip
          - pip:
              # This should eventually move to rapids-dask-dependency
              - &dask_expr_tip git+https://github.com/dask-contrib/dask-expr.git@b588a9e15e90e0567061664ffc01374786686e20
      - output_types: [requirements, pyproject]
        packages:
          - *dask_expr_tip
```

Essentially, this is just saying that the conda file should include `pip` and the `pip: [git+...]` mapping, while the pyproject and requirements files should only have the `git+...` link itself.

Comment (rjzamora) [reply dependencies.yaml]: Aha. That makes a lot of sense - Thanks @charlesbluca !

File: dependencies.yaml
Comment (charlesbluca) [inline dependencies.yaml:672]: Looking at the errors, seems like the pyproject requirement needs to be PEP 508 compliant, which I _think_ should look like this:

```suggestion
      - output_types: requirements
        packages:
          - *dask_expr_tip
      - output_types: pyproject
        packages:
          - dask-expr@git+https://github.com/dask-contrib/dask-expr.git@b588a9e15e90e0567061664ffc01374786686e20
```

File: dependencies.yaml
Comment (charlesbluca) [inline dependencies.yaml:675]: It would be nice to use `dask_expr_tip` here to cut down on duplication, but have no idea if something like

```suggestion
          - dask-expr @ *dask_expr_tip
```

Would work here as I'm not the most familiar with `rapids-dependency-file-generator`

Comment (rjzamora) [reply dependencies.yaml:675]: Sounds fine to me. As you know, I'm just deferring to your advice for these kinds of changes :)

Another background note: I'm expecting dask-expr to be copied entirely into dask/dask proper before the "query-planning" default is changed. When that happens, we should be able to remove all this stuff.

Comment (rjzamora) [reply dependencies.yaml:675]: Hmm - I'm not having much luck with this. Not sure if we _can_ avoid the duplication in this case.

File: python/dask_cudf/dask_cudf/backends.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/backends.py:502]: TODO: Revert this change (was debugging a separate cudf bug)

File: python/dask_cudf/dask_cudf/expr/__init__.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/expr/__init__.py:13]: This name means something different than the `DASK_EXPR_ENABLED` variable in `dask.dataframe`. Should probably be removed from `dask_cudf`, since it isn't used anyway.

File: dependencies.yaml
Comment (rjzamora) [inline dependencies.yaml:652]: Latest version of dask-expr supported for dask 2024.1.1

File: python/cudf/cudf/core/frame.py
Comment (rjzamora) [inline python/cudf/cudf/core/frame.py:1961]: **Note**: After some recent tokenization fixes in dask, we now need to "stringify" this information in order for the token to be deterministic.

File: python/dask_cudf/dask_cudf/__init__.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/__init__.py:11]: Top-level imports must pass through `dask.dataframe` in order for the `"query-planning"` to be picked up correctly. This should not break/change "legacy" behavior.

File: python/dask_cudf/dask_cudf/expr/__init__.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/expr/__init__.py:6]: We need to know at import time if dask-exp is "enabled" (i.e. query planning is "on")

File: python/dask_cudf/dask_cudf/expr/_collection.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/expr/_collection.py:26]: Dask-expr's `var` implementation doesn't work for cudf. May be able to upstream a "fix" for later dask-expr versions.

File: python/dask_cudf/dask_cudf/backends.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/backends.py:639]: There is nothing "new" here. We have already defined/registered a `DataFrameBackendEntrypoint` class for legacy dask-dataframe collections. Now we need to define/register a new one class for new expr-based collections. The difinitions will look very similar, but the classes need to be different.

File: python/dask_cudf/dask_cudf/core.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/core.py:723]: Honestly haven't had time to understand this.

File: python/dask_cudf/dask_cudf/expr/_expr.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/expr/_expr.py:1]: I would consider the logic in this class to be the most "fragile". We are literally patching `Expr` sub-classes to do things that will work for cudf-backed data. By "surgically" patching specific classes, we dramatically reduce the amount of code we need in dask-cudf.

File: python/dask_cudf/dask_cudf/expr/_expr.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/expr/_expr.py:26]: Dask-expr's `CumulativeBlockwise` assumes positional arguments that are true for pandas, but not cudf.

File: python/dask_cudf/dask_cudf/expr/_expr.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/expr/_expr.py:85]: Avoiding `get_group` (https://github.com/rapidsai/cudf/issues/14955)

File: python/dask_cudf/dask_cudf/expr/_expr.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/expr/_expr.py:104]: `DecomposableCudfGroupbyAggregation` is essentially dask-cudf's "optimized" groupby aggregation. Note that this does not seem to be any faster than dask-expr's algorithm any longer (perhaps groupby-apply behavior is much faster in cudf now?). However, we still use a special code path for now to support aggregations like "collect". (tldr: We may be able to remove the custom dask-cudf groupby-aggregation code path soon)

File: python/dask_cudf/dask_cudf/expr/_expr.py
Comment (wence-) [reply python/dask_cudf/dask_cudf/expr/_expr.py:85]: https://github.com/rapidsai/cudf/pull/15143

**Review by wence- — COMMENTED** [review]
Partial pass

File: dependencies.yaml
Comment (rjzamora) [reply dependencies.yaml:652]: That is indeed my preference: https://github.com/rapidsai/rapids-dask-dependency/issues/23

File: python/cudf/cudf/core/index.py
Comment (rjzamora) [reply python/cudf/cudf/core/index.py:963]: Thats a good question - Need to follow up on this.

File: python/dask_cudf/dask_cudf/expr/_collection.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/expr/_collection.py:25]: Yeah, I probably like that better - If I remember correctly, this had more than a var fix at one point, but we might as well keep the mixins separate for clarity.

File: python/dask_cudf/dask_cudf/expr/_expr.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/expr/_expr.py:26]: Yes - You've probably noticed that many of these "patches" can also be replced with fixes/features in cudf. It probably makes sense to open an issue and link it in a comment for each of these patches so it is easier to figure out when a patch can be removed.

File: python/dask_cudf/dask_cudf/expr/_collection.py
Comment (wence-) [reply python/dask_cudf/dask_cudf/expr/_collection.py:26]: OK, so this is because column-indexing of a DataFrame doesn't handle device-side objects, which seems like a missing cudf feature

**Review by wence- — COMMENTED** [review]
A few more comments. I think this overall looks good, but I would prefer to open issues and assign them to fix the various cudf/pandas incompatibilities we're working around here before merging

File: python/dask_cudf/dask_cudf/expr/_expr.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/expr/_expr.py:47]: https://github.com/dask-contrib/dask-expr/blob/25d4717f47a2cc74339047f774792f627d18b858/dask_expr/_cumulative.py#L63
?

File: python/dask_cudf/dask_cudf/tests/test_applymap.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/tests/test_applymap.py:28]: Yes `DataFrame.applymap` is deprecated since Pandas 2.1.0, so makes more sense to test `DataFrame.map` moving forward.

File: python/dask_cudf/dask_cudf/expr/_expr.py
Comment (wence-) [reply python/dask_cudf/dask_cudf/expr/_expr.py:47]: Oh, I'm blind

File: python/dask_cudf/dask_cudf/io/tests/test_parquet.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/io/tests/test_parquet.py:298]: This is sort of discussed in https://github.com/dask-contrib/dask-expr/issues/800, and was fixed by https://github.com/dask-contrib/dask-expr/pull/798 (which is in `dask-expr>=0.4.1`). The general problem was that the expression caching was a bit too aggressive, and so we didn't bother to re-read a modified dataset with the same path. I suppose I can link the issue, and add a comment that the `overwrite=True` can be removed for `dask-expr>=0.4.1`?

File: python/cudf/cudf/core/frame.py
Comment (rjzamora) [reply python/cudf/cudf/core/frame.py:1961]: Okay, looks like tokenizing types is not the problem, so I will roll back part of this change. As far as I can tell, https://github.com/dask/dask/pull/10904 tweaked the behavior of nested `dict` normalization, and what we really want is to avoid normalizing the `_dtypes` within `__dask_tokenize__` like this. Can't say I completely understand what was failing in `test_dispatch.py::test_deterministic_tokenize` without this "fix".

File: python/dask_cudf/dask_cudf/expr/_expr.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/expr/_expr.py:33]: Will link this: https://github.com/rapidsai/cudf/issues/15177

File: python/dask_cudf/dask_cudf/expr/_expr.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/expr/_expr.py:83]: Will definitely remove this and re-test once 15143 is merged, thanks!

File: python/dask_cudf/dask_cudf/expr/_groupby.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/expr/_groupby.py:48]: Added a comment linking https://github.com/rapidsai/cudf/issues/15173 above.

File: python/dask_cudf/dask_cudf/groupby.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/groupby.py:808]: It's all a matter of conforming the "optimized-for-cudf" groupby aggregation to the existing dask-expr classes. The "decomposed" groupby aggregation was requiring us to flatten columns earlier. However, now that custom groupby aggregations have been removed fro this PR, I will roll back these changes.

File: python/dask_cudf/dask_cudf/expr/_collection.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/expr/_collection.py:28]: Linked an issue for this.

File: python/dask_cudf/dask_cudf/expr/_expr.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/expr/_expr.py:14]: linked relevant issue.

**Comment by rjzamora** [comment]
I'm not sure I understand why CI is now "green" but there are still 3.11 test failures for dask_cudf and dask_cudf + dask-expr (https://github.com/rapidsai/cudf/actions/runs/8090360941/job/22108518925)

**Comment by bdice** [comment]
> I'm not sure I understand why CI is now "green" but there are still 3.11 test failures for dask_cudf and dask_cudf + dask-expr (https://github.com/rapidsai/cudf/actions/runs/8090360941/job/22108518925)

You can ignore Python 3.11 for now. We are forcing those jobs to succeed so they don’t block CI. See #15111 and https://github.com/rapidsai/shared-workflows/pull/182.

**Comment by rjzamora** [comment]
>You can ignore Python 3.11 for now.

Ah, okay. Thanks for the info @bdice !

**Comment by rjzamora** [comment]
@wence- I rolled back changes that merged in https://github.com/rapidsai/cudf/pull/15143 in case it makes it possible to merge this sooner. (Does the memcheck + dask-cudf bug only shows up with those particular changes?)

**Comment by wence-** [comment]
> @wence- I rolled back changes that merged in #15143 in case it makes it possible to merge this sooner. (Does the memcheck + dask-cudf bug only shows up with those particular changes?)

No clue sorry, I haven't gone through all of the other PRs.

File: python/dask_cudf/dask_cudf/core.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/core.py:723]: This is because the doc-string is currently missing in dask-expr

**Review by wence- — APPROVED** [review]
I think this is a good step. I would prefer if we didn't have to patch so much, but thanks for opening all the issues!

File: python/dask_cudf/dask_cudf/expr/_collection.py
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/expr/_collection.py:72]: I think this is mostly guarding against a `pd.Grouper` argument.

**Review by raydouglass — APPROVED** [review]
`ops-codeowners` changes look good

**Pull request merged**
